Add comprehensive documentation for policy table structure (#50)
smartdevicelink/sdl_overview_guides: docs/Policies/App Policies/index.md

"app_policies": {
    "123abc": {
        "keep_context": false,
        "steal_focus": false,
        "priority": "NONE",
        "default_hmi": "NONE",
        "groups": ["Base-4"],
        "RequestType": [],
        "RequestSubType": [],
        "nicknames": ["Example Application"],
        "AppHMIType": ["NAVIGATION", "MEDIA"]
    },
    ...
}
```

### groups

This field defines a list of functional groups which should be made available to this application, populated with a list of keys from the `functional_groupings` section (ex. if this list contains "Base-4", all permissions defined in the "Base-4" functional group will be assigned to this application).

### preconsented_groups

This field defines additional groups which are assigned to this application, with the difference that the user is expected to have consented to using these permissions separately from the HMI (via a mobile popup, etc.). Any groups defined in this list do not require consent from the user, even if they have a `user_consent_prompt` defined. Only applies when Core is built with `EXTERNAL_PROPRIETARY` policy mode, otherwise these permissions are simply added to the normal `groups` list.

### nicknames

> Defines a list of potential appNames which an application is allowed to register with. If the appName field in the app's `RegisterAppInterface` does not match one of the values in this list, the message will be rejected. If omitted, no restrictions will be applied to appName.

Reviewer: ```suggestion
This field defines a list of potential appNames which an application is allowed to register with. If the appName field in the app's `RegisterAppInterface` does not match one of the values in this list, the message will be rejected. If omitted, no restrictions will be applied to appName.
```

!!! note
This field is mandatory when defining policies for a cloud application, in which case the first nickname will be used in the application list prior to app registration.
!!!

### AppHMIType

> Defines a list of AppHMITypes which an application is allowed to register with. If omitted, all AppHMITypes will be allowed.

Reviewer: ```suggestion
This field defines a list of AppHMITypes which an application is allowed to register with. If omitted, all AppHMITypes will be allowed.
```

### keep_context

> This field is a flag which allows or disallows an application to use soft buttons with the `KEEP_CONTEXT` action. Only applies when Core is built with `EXTERNAL_PROPRIETARY` policy mode, otherwise this permission is always allowed.

Reviewer: ```suggestion
This field is a flag which allows or disallows an application to use soft buttons with the `KEEP_CONTEXT` action. It only applies when Core is built with `EXTERNAL_PROPRIETARY` policy mode, otherwise this permission is always allowed.
```

### steal_focus

> This field is a flag which allows or disallows an application to use soft buttons with the `STEAL_FOCUS` action. Only applies when Core is built with `EXTERNAL_PROPRIETARY` policy mode, otherwise this permission is always allowed.

Reviewer: ```suggestion
This field is a flag which allows or disallows an application to use soft buttons with the `STEAL_FOCUS` action. It only applies when Core is built with `EXTERNAL_PROPRIETARY` policy mode, otherwise this permission is always allowed.
```

### priority

String value tied to the `notifications_per_minute_by_priority` section of `module_config`. By assigning a priority value from this list, the application is allowed the number of background alerts which is specified by that category (see [notifications_per_minute_by_priority](../policy-fields#notifications_per_minute_by_priority)).

### default_hmi

> Defines the default HMI level that the application will be initialized to when it is registered for the first time. Only applies when Core is built with `EXTERNAL_PROPRIETARY` policy mode, otherwise the default HMI level is always `NONE`. 

Reviewer: ```suggestion
This field defines the default HMI level that the application will be initialized to when it is registered for the first time. It only applies when Core is built with `EXTERNAL_PROPRIETARY` policy mode, otherwise the default HMI level is always `NONE`. 
```

### RequestType

> Defines a list of possible values for `requestType` that can be used in `OnSystemRequest` and `SystemRequest` RPCs when communicating with this app.

Reviewer: ```suggestion
This field defines a list of possible values for `requestType` that can be used in `OnSystemRequest` and `SystemRequest` RPCs when communicating with this app.
```

### RequestSubType

> Defines a list of possible values for `requestSubType` that can be used in `OnSystemRequest` and `SystemRequest` RPCs when communicating with this app.

Reviewer: ```suggestion
Defines a list of possible values for `requestSubType` that can be used in `OnSystemRequest` and `SystemRequest` RPCs when communicating with this app.
```
```suggestion
This field defines a list of possible values for `requestSubType` that can be used in `OnSystemRequest` and `SystemRequest` RPCs when communicating with this app.
```

        "certificate": "-----BEGIN CERTIFICATE-----\n...",
        "encryption_required": true
    },
    ...
}
```

### certificate

A string field containing a certificate for establishing a secure connection with the application. Currently only used for Cloud Applications when establishing a secure websocket connection.

### encryption_required

This flag which can be used to override `encryption_required` on the functional group level. If `false`, then this application will not be required to encrypt any of the RPCs which it has permissions to, even if the associated functional group has `encryption_required` set to `true`. Defaults to `true` if omitted.

## Remote Control Fields

!!! note
These permissions will only take effect if the application registers with the `REMOTE_CONTROL` AppHMIType
!!!

### Example Entry

```json
"app_policies": {
    "123abc": {
        "keep_context": false,
        "steal_focus": false,
        "priority": "NONE",
        "default_hmi": "NONE",
        "groups": ["Base-4", "RemoteControl"],
        "RequestType": [],
        "RequestSubType": [],
        "moduleType": ["RADIO", "CLIMATE"]
    },
    ...
}
```

### moduleType

> Defines a list of Remote Control module types (defined in the `ModuleType` enum in the [Mobile API](https://github.com/smartdevicelink/rpc_spec#moduletype)) which are accessible by the app. If empty, all module types will be allowed. If omitted, no module types will be allowed.

Reviewer: ```suggestion
This field defines a list of Remote Control module types (defined in the `ModuleType` enum in the [Mobile API](https://github.com/smartdevicelink/rpc_spec#moduletype)) which are accessible by the app. If empty, all module types will be allowed. If omitted, no module types will be allowed.
```

## App Service Fields

> These fields are used if the application is either be a provider or consumer of app service data. See our [App Service Provider Guide](https://smartdevicelink.com/en/guides/core/app-service-provider-guide/) for more information on developing an application with these capabilities.

Reviewer: ```suggestion
These fields are used if the application is either a provider or consumer of app service data. See our [App Service Provider Guide](https://smartdevicelink.com/en/guides/core/app-service-provider-guide/) for more information on developing an application with these capabilities.
```

* `service_names` - Defines a list of potential service names that an app can publish with this service type. 
    * If the `serviceName` field in the app's `PublishAppService.appServiceManifest` does not match one of the values in this list, the message will be rejected.
    * If omitted, no restrictions will be applied to the `serviceName` parameter.
* `handled_rpcs` - This field defines a list of permissions for RPCs that an app service could intercept via the [RPC passing](https://smartdevicelink.com/en/guides/core/app-service-provider-guide/#rpc-passing) feature. 
    * If the `handledRPCs` field in the app's `PublishAppService.appServiceManifest` includes a function ID that is not in this list, the message will be rejected. 
    * This field is mandatory, but can be empty.
    * Subfields:
        * `function_id` - The [function ID](https://github.com/smartdevicelink/sdl_core/blob/master/src/components/interfaces/MOBILE_API.xml) of the potential handled RPC. Can be a function ID which was introduced after Core's local RPC spec version.

### allow_unknown_rpc_passthrough

> This flag can be used to allow an application to pass RPCs which are unknown to Core (functions which were defined after Core's local API version) to an app service provider. Otherwise, such messages from this application will be rejected. Defaults to `false`.

Reviewer: >Defaults to `false`.

Can you please be more specific here?  For example, "This flag defaults to `false`."

# App Policies

> The app policies section defines permissions that each application is granted in the SDL environment. This is where you would change the default permissions for each application, or add rules for specific applications. Each key in this object is associated with the `fullAppID` field for an application (or `appID` if `UseFullAppID = false` in `smartDeviceLink.ini`), and each value is a struct which contains several fields to control permissions for this app. In addition, several reserved keys exist for this section:

Reviewer 2: I was about to comment and say that 2 is not several in my book but then I saw the TODO, 3 is on the edge of several, up to you.

* `default` - The set of permissions which are assigned to an app if it does not have its own entry in the app policies section.
* `pre_DataConsent` - Only applies when Core is built with `EXTERNAL_PROPRIETARY` policy mode. The set of permissions which are assigned to an app if the device is not given SDL permissions by the user.

> <!--* `device` - [TODO: find the exact purpose for this key]-->

Reviewer 2: TODO ;)

### steal_focus

This field is a flag which allows or disallows an application to use soft buttons with the `STEAL_FOCUS` action. Only applies when Core is built with `EXTERNAL_PROPRIETARY` policy mode, otherwise this permission is always allowed.

### priority

String value tied to the `notifications_per_minute_by_priority` section of `module_config`. By assigning a priority value from this list, the application is allowed the number of background alerts which is specified by that category (see [notifications_per_minute_by_priority](../policy-fields#notifications_per_minute_by_priority)).

### default_hmi

Defines the default HMI level that the application will be initialized to when it is registered for the first time. Only applies when Core is built with `EXTERNAL_PROPRIETARY` policy mode, otherwise the default HMI level is always `NONE`. 

### RequestType

Defines a list of possible values for `requestType` that can be used in `OnSystemRequest` and `SystemRequest` RPCs when communicating with this app.

### RequestSubType

Defines a list of possible values for `requestSubType` that can be used in `OnSystemRequest` and `SystemRequest` RPCs when communicating with this app.

> ### memory_kb

Reviewer 2: im thinking maybe this should be removed from the guide if we have no plan to implement it in the future

Author: Might consider keeping it just for completeness, but commenting it out so that it doesn't appear on the developer portal.

Unused, see https://github.com/smartdevicelink/sdl_core/issues/3127.

### heart_beat_timeout_ms

> Heartbeat timeout for this application, deprecated in [Protocol Version 4](https://github.com/smartdevicelink/protocol_spec#45-heartbeat).

Reviewer 2: It might be helpful to have a date here as well, so people can quickly tell if it was deprecated before/after they started using Core

Author: In practice, this has been the case almost since the project was made open source, but I can link to Release 4.0.0, where protocol version 4 was introduced.

> ### Example Entry

Reviewer 2: i think seeing the examples at the beginning of each section would be more helpful than after reading about each property

!!!

### moduleType

Defines a list of Remote Control module types (defined in the `ModuleType` enum in the [Mobile API](https://github.com/smartdevicelink/rpc_spec#moduletype)) which are accessible by the app. If empty, all module types will be allowed. If omitted, no module types will be allowed.

### Example Entry

```json
"app_policies": {
    "123abc": {
        "keep_context": false,
        "steal_focus": false,
        "priority": "NONE",
        "default_hmi": "NONE",
        "groups": ["Base-4", "RemoteControl"],
        "RequestType": [],
        "RequestSubType": [],
        "moduleType": ["RADIO", "CLIMATE"]
    },
    ...
}
```

## App Service Fields

These fields are used if the application is either be a provider or consumer of app service data. See our [App Service Provider Guide](https://smartdevicelink.com/en/guides/core/app-service-provider-guide/) for more information on developing an application with these capabilities.

### app_services

This field defines the properties of the app services that this application is allowed to publish. Each key within this object corresponds to an app service type (ex. "MEDIA").

#### service_names

> Defines a list of potential serviceNames that an app can publish with this service type. If the serviceName field in the app's `PublishAppService.appServiceManifest` does not match one of the values in this list, the message will be rejected. If omitted, no restrictions will be applied to serviceName.

Reviewer 2: ```suggestion
Defines a list of potential service names that an app can publish with this service type. If the `serviceName` field in the app's `PublishAppService.appServiceManifest` does not match one of the values in this list, the message will be rejected. If omitted, no restrictions will be applied to the service name.
```

I feel like it should be 'service names' in the first line, maybe in backticks the second time, and then with the space again the last time.

#### handled_rpcs

This field defines a list of permissions for RPCs that an app service could intercept via the [RPC passing](https://smartdevicelink.com/en/guides/core/app-service-provider-guide/#rpc-passing) feature. If the handledRPCs field in the app's `PublishAppService.appServiceManifest` includes a function ID that is not in this list, the message will be rejected. This field is mandatory, but can be empty.

##### function_id

> The [function ID](https://github.com/smartdevicelink/sdl_core/blob/master/src/components/interfaces/MOBILE_API.xml) of the potential handled RPC. Can be an function ID which was introduced after Core's local RPC spec version.

Reviewer 2: ```suggestion
The [function ID](https://github.com/smartdevicelink/sdl_core/blob/master/src/components/interfaces/MOBILE_API.xml) of the potential handled RPC. Can be a function ID which was introduced after Core's local RPC spec version.
```

`an function ID` -> `a function ID`

        "RequestType": [],
        "RequestSubType": [],
        "app_services": {
            "MEDIA": {
                "service_names": ["SDL Music", "App Music"],
                "handled_rpcs": [{"function_id": 41}]
            },
            "NAVIGATION": {
                "service_names": ["SDL Navigation", "App Navigation"],
                "handled_rpcs": [{"function_id": 39}]
            },
            "WEATHER": {
                "service_names": ["SDL Weather", "App Weather"],
                "handled_rpcs": []
            }
        },
        "allow_unknown_rpc_passthrough": true
    },
    ...
}
```

## Cloud Application Fields

These fields are used for applications which use a cloud transport and require Core to initiate the connection. Each of these fields can be modified by a permitted app via the `SetCloudAppProperties` RPC.

### enabled

This flag is used to determine if the cloud application should appear in the app list in the HMI. If `true`, the application will appear in the app list as a cloud application. Selecting this application from the list will cause Core to connect to and register the application.

### hybrid_app_preference

> This option is used to determine the default behavior if the same app is available via a mobile device and over the cloud (determined by a duplicate appName). The possible values for this option are `MOBILE`, `CLOUD`, or `BOTH`. 

Reviewer 2: ```suggestion
This option is used to determine the default behavior if the same app is available via a mobile device and over the cloud (determined by a duplicate app name). The possible values for this option are `MOBILE`, `CLOUD`, or `BOTH`. 
```
fix camelCase in documentation, alternatively could add backticks

## Cloud Application Fields

These fields are used for applications which use a cloud transport and require Core to initiate the connection. Each of these fields can be modified by a permitted app via the `SetCloudAppProperties` RPC.

### Example Entry

```json
"app_policies": {
    "123abc": {
        "keep_context": false,
        "steal_focus": false,
        "priority": "NONE",
        "default_hmi": "NONE",
        "groups": ["Base-4", "RemoteControl"],
        "RequestType": [],
        "RequestSubType": [],
        "endpoint": "ws://smartdevicelink.com:8888/path/",
        "enabled": true,
        "cloud_transport_type": "WS",
        "auth_token": "1234567890",
        "icon_url": "http://i.imgur.com/TgkvOIZ.png"
    },
    ...
}
```

### enabled

This flag is used to determine if the cloud application should appear in the app list in the HMI. If `true`, the application will appear in the app list as a cloud application. Selecting this application from the list will cause Core to connect to and register the application.

### hybrid_app_preference

> This option is used to determine the default behavior if the same app is available via a mobile device and over the cloud (determined by a duplicate `appName` search). The possible values for this option are `MOBILE`, `CLOUD`, or `BOTH`. 

Reviewer 2: I think this statement sounds better without the `search` after appName
